soda/core/soda/execution/data_source.py

```diff
@@ -488,26 +487,33 @@ def profiling_sql_aggregates_text(self, table_name: str, column_name: str) -> st
         )
 
     def histogram_sql_and_boundaries(
-        self, table_name: str, column_name: str, min: int | float, max: int | float
+        self,
+        table_name: str,
+        column_name: str,
+        min_value: int | float,
+        max_value: int | float,
+        n_distinct: int,
+        column_type: str,
     ) -> tuple[str | None, list[int | float]]:
         # TODO: make configurable or derive dynamically based on data quantiles etc.
-        number_of_bins: int = 20
+        max_n_bins = 20
+        number_of_bins: int = max(1, min(n_distinct, max_n_bins))
+        number_of_intervals: int = number_of_bins - 1
 
-        if not min < max:
+        if min_value >= max_value:
             self.logs.warning(
-                f"Min of {column_name} on table: {table_name} must be smaller than max value. Min is {min}, and max is {max}"
+                f"Min of {column_name} on table: {table_name} must be smaller than max value. Min is {minimum}, and max is {maximum}"
             )
             return None, []
 
-        min_value = floor(min * 1000) / 1000
-        max_value = ceil(max * 1000) / 1000
-        bin_width = (max_value - min_value) / number_of_bins
+        bin_width = (max_value - min_value) / number_of_intervals
+
+        if bin_width.is_integer() and column_type == "integer":
+            bin_width = int(bin_width)
+            min_value = int(min_value)
+            max_value = int(max_value)
 
-        boundary_start = min_value
-        bins_list = [min_value]
-        for _ in range(0, number_of_bins):
-            boundary_start += bin_width
-            bins_list.append(round(boundary_start, 3))
+        bins_list = [round(min_value + i * bin_width, 2) for i in range(0, number_of_bins)]
```

Comment:
Tiny difference with previous code regarding functionality because `min_value` did not use to be rounded and it is now

Reply:
isn't it same in both cases we sum it first and then round it?

Reply:
hmm indeed good catch, then the first min_value was rounded in the upper lines and now it's rounded in the list.

Reply:
Btw, wouldn't it make more sense to deal with the formatting in the front-end? Is it correct that we're rounding to two decimals because of the way the bins will displayed?

Reply:
Not sure if any of this matters though haha, probably not important for now right?

Reply:
yes, that's true. Personally, to have more control I would prefer to handle it on soda-core. Wdyt @bastienboutonnet ?

Reply:
Front end most definitely won't apply any domain logic. Cloud only receives data and doesn't perform any other kind of logic on it. So indeed this should be something soda-core is responsible for.